### DISCUSSION TITLE
intermediate saving of data and schema attributes

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#158] [Design & Development]
In reviewing the current yaml file for the mspass schema I realized a fundamental hole:  we need one or more mechanisms to provide a unique tag in a wf collection defining an intermediate save and/or a save of the final processed data for a workflow.  There is nothing right now.  Without something it would be hard to separate raw from processed data with any simple query.   I see two issues:
1.  What key-value pair do we use to tag a specific save?
2.  How should that be set in the api?

The two issues are not orthogonal.  I think the answer to 1 is pretty simple.  We just need to decide on a sensible name key and what type it should be.  My suggestion is use "save_tag" or "save_name", but open to any suggestion.  I think the type should just be a string.  That way it is pretty flexible.  

To me a fundamental issue is if we use the as yet incompletely designed global history manager to manage this or use a purely manual method?   I would suggest both be supported with this modification to waveform save method.  that is, two new args:
1.  save_tag=None - manual method.   When defined use the value of this arg as the tag for the save. 
2. get_tag_from_manager=True.  Default would get the name from the global history manager - assuming our design makes that feasible.  

The arg names are maybe too verbose, but they are descriptive.  I think the key discussion issue is if we want more than one method to set a tag for a save at a particular state.

**Comment by wangyinz** [comment]
That's a good point. I initially thought we could advise the user to save to a different collection or even database to distinguish those data, but I think this tag feature is a more convenient addition. 

I would call this new key `data_tag` because the tagging is not necessarily a concept unique to "save". Also, I don't think it necessary to add this key to our database schema if we already defined it as an argument in the database API - schema doesn't have the control on the API's behavior for this key anyway. Also, this `data_tag=None` argument should be added to read, save, and update methods, and in the read method the default behavior is to not check for data_tag if not defined, but the read will ignore any documents of mismatched tags.

The `get_tag_from_manager` argument is not needed for now. Before implementing that, we need to think carefully on the behavior of the manager. I am not sure how a global history manager can generate sensible tags automatically at the moment. Will need to revise the design for sure.
